srcjs/input_binding_radio.js

```diff
@@ -5,10 +5,25 @@ $.extend(radioInputBinding, {
   },
   getValue: function(el) {
     // Select the radio objects that have name equal to the grouping div's id
-    return $('input:radio[name="' + $escape(el.id) + '"]:checked').val();
+    var value = $('input:radio[name="' + $escape(el.id) + '"]:checked').val();
+    // If none are checked, the input will return null (it's the default on load,
+    // but it wasn't emptied when calling updateRadioButtons with character(0)
+    if (value) {
+      return value;
+    } else {
+      return null;
+    }
   },
   setValue: function(el, value) {
-    $('input:radio[name="' + $escape(el.id) + '"][value="' + $escape(value) + '"]').prop('checked', true);
+    if (value === undefined) {
+      return;
+    } else if (value.length === 0) {
```

Comment:
This check feels weird to me. The behavior we want is that if `value` is `[]`, it does one thing, and if it's a string, it does another.

`value.length === 0` will be true if `value` is `[]`, but it will also be true if if `value` is `""`, because `.length` is also defined for strings.

Now, I don't think it's possible for `updateRadioButtons()` to send a value of `""`, so it won't cause problems, but this code is still a bit confusing.

It would be clearer if this first checked the type (with `$.isArray()`) and then the length.

Reply:
Hey @wch, 

Ok, let me know if I get this straight: 

The browser receives a values, it can be: 

- an empty array, when `updateRadioButtons(session, "radio", selected = character(0))`, then we uncheck all the radioButtons
- a single value, when `updateRadioButtons(session, "radio", selected = letters[1])`, then we update the selected radiobutton
- an array of value, when doing `updateRadioButtons(session, "radio", selected = letters[1:5])` then shiny fails in the browser with `Uncaught TypeError: e.replace is not a function` => see https://github.com/rstudio/shiny/issues/3065
- an `""`, but I'm not sure when this happens? In any case, this might be handled as if `selected = character(0)`?

Reply:
I think it's reasonable to just do the following:

```js
if ($.isArray(value) && value.length === 0) {
  // This is the character(0) case 
  ...
} else {
  ...
}
```

This isn't so different from what you have right now. The differences are:

* No `undefined` check. If an `undefined` gets here, there's probably a bug somewhere, and it would be helpful for it to throw an error so that the bug can be more easily found.
* It just adds an array test to the `value.length===0` check. This is because, without the array check, it's confusing that `value.length` could be 0 if it's an array OR a string, and someone reading the code could be misled by it. When I looked at the code, I initially thought it was testing for `""`, because elsewhere in the function it's expected to be a string. Adding the `$.isArray()` test makes it unambiguous what that if-statement is really looking for.

In my limited testing, I wasn't able to get `updateRadioButtons` to send a `""` as the value. I think it gets removed on the R side first. I don't think we need a specific check for it.

Reply:
@wch I've put this new test version in the last commit :)